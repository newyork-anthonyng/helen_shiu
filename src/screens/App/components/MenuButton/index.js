import React, { Children, cloneElement } from 'react';
import T from 'prop-types';
import noop from '../../utilities/noop';

const MenuButton = ({ children, onClick, onKeyDown }) => {
  return cloneElement(Children.only(children), {
    onClick,
    onKeyDown,
    tabIndex: 0,
  });
};

MenuButton.defaultProps = {
  onClick: noop,
  onKeyDown: noop,
};

MenuButton.propTypes = {
  onClick: T.func.isRequired,
  onKeyDown: T.func.isRequired,
};

export default MenuButton;
