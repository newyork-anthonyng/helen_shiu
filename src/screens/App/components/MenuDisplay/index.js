import React, { Component, Children } from 'react';
import T from 'prop-types';

class MenuDisplay extends Component {
  componentDidMount() {
    let nextChildIndex = 0;
    if (this.props.shouldFocusOnLastElement) {
      nextChildIndex = Children.count(this.props.children) - 1;
    }

    this.refs[`child-${nextChildIndex}`].focus();
  }

  handleKeyDown = (index) => {
    return (e) => {
      let nextFocusableChildIndex;
      const childrenCount = Children.count(this.props.children);

      switch (e.keyCode) {
        case 38: // UP
          nextFocusableChildIndex = index - 1;
          if (nextFocusableChildIndex < 0) {
            nextFocusableChildIndex = childrenCount - 1;
          }
          break;
        case 40: // DOWN
          nextFocusableChildIndex = index + 1;
          if (nextFocusableChildIndex >= childrenCount) {
            nextFocusableChildIndex = 0;
          }
          break;
      }

      if (nextFocusableChildIndex !== undefined) {
        this.refs[`child-${nextFocusableChildIndex}`].focus();
      }
    }
  }

  renderChildren = () => {
    return Children.map(this.props.children, (child, index) => {
      return React.cloneElement(child, {
        ref: `child-${index}`,
        tabIndex: -1,
        onKeyDown: this.handleKeyDown(index),
      });
    });
  }

  render() {
    return (
      <div>
        {this.renderChildren()}
      </div>
    );
  }
}

MenuDisplay.defaultProps = {
  shouldFocusOnLastElement: false,
};

MenuDisplay.propTypes = {
  shouldFocusOnLastElement: T.bool,
};

export default MenuDisplay;
