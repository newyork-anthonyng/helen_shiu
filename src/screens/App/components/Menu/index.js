import React, { Component, Children, cloneElement } from 'react';
import MenuDisplay from '../MenuDisplay';
import MenuButton from '../MenuButton';

class Menu extends Component {
  constructor() {
    super();

    this.state = {
      isMenuVisible: false,
      shouldFocusOnLastElement: false,
    };
  }

  handleMenuButtonClick = () => {
    this.setState(prevState => ({
      isMenuVisible: !prevState.isMenuVisible,
      shouldFocusOnLastElement: false,
    }));
  }

  handleMenuButtonKeyDown = (e) => {
    // 38 => UP
    // 40 => DOWN
    // 13 => ENTER
    // 32 => SPACE_BAR
    if (!/(38|40|32|13)/.test(e.keyCode)) return;

    const shouldFocusOnLastElement = e.keyCode === 38;

    this.setState(prevState => ({
      isMenuVisible: !prevState.isMenuVisible,
      shouldFocusOnLastElement,
    }));
  }

  renderChildren = () => {
    return Children.map(this.props.children, child => {
      if (child.type === MenuDisplay) {
        if (this.state.isMenuVisible) {
          return cloneElement(child, {
            shouldFocusOnLastElement: this.state.shouldFocusOnLastElement,
          });
        }
        return null;
      } else if (child.type === MenuButton) {
        return cloneElement(child, {
          onClick: this.handleMenuButtonClick,
          onKeyDown: this.handleMenuButtonKeyDown,
        });
      }
      return child;
    });
  }

  render() {
    return (
      <div>
        {this.renderChildren()}
      </div>
    );
  }
}

export default Menu;
