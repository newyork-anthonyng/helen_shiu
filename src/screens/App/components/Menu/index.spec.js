import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Menu from './';
import MenuDisplay from '../MenuDisplay';
import MenuButton from '../MenuButton';

it('should return correctly', () => {
  const wrapper = mount(
    <Menu>
      <MenuButton><a>Toggle Menu</a></MenuButton>
      <MenuDisplay>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    </Menu>
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

describe('It should toggle Menu Display', () => {
  it('when clicking button', () => {
    const wrapper = mount(
      <Menu>
        <MenuButton><a>Toggle Menu</a></MenuButton>
        <MenuDisplay>
          <h1>First</h1>
          <h1>Second</h1>
          <h1>Third</h1>
        </MenuDisplay>
      </Menu>
    );

    const menuButton = wrapper.find(MenuButton);
    menuButton.simulate('click');
    expect(wrapper.find(MenuDisplay).length).toEqual(1);
    expect(document.activeElement).toEqual(wrapper.find('h1').get(0));

    menuButton.simulate('click');
    expect(wrapper.find(MenuDisplay).length).toEqual(0);
  });

  it('when pressing Space Bar', () => {
    const wrapper = mount(
      <Menu>
        <MenuButton><a>Toggle Menu</a></MenuButton>
        <MenuDisplay>
          <h1>First</h1>
          <h1>Second</h1>
          <h1>Third</h1>
        </MenuDisplay>
      </Menu>
    );
    const SPACE_BAR = 32;

    const menuButton = wrapper.find(MenuButton);
    menuButton.simulate('keydown', { keyCode: SPACE_BAR });
    expect(wrapper.find(MenuDisplay).length).toEqual(1);
    expect(document.activeElement).toEqual(wrapper.find('h1').get(0));

    menuButton.simulate('keydown', { keyCode: SPACE_BAR });
    expect(wrapper.find(MenuDisplay).length).toEqual(0);
  });

  it('when pressing Enter', () => {
    const wrapper = mount(
      <Menu>
        <MenuButton><a>Toggle Menu</a></MenuButton>
        <MenuDisplay>
          <h1>First</h1>
          <h1>Second</h1>
          <h1>Third</h1>
        </MenuDisplay>
      </Menu>
    );
    const ENTER_KEY = 13;

    const menuButton = wrapper.find(MenuButton);
    menuButton.simulate('keydown', { keyCode: ENTER_KEY });
    expect(wrapper.find(MenuDisplay).length).toEqual(1);
    expect(document.activeElement).toEqual(wrapper.find('h1').get(0));

    menuButton.simulate('keydown', { keyCode: ENTER_KEY });
    expect(wrapper.find(MenuDisplay).length).toEqual(0);
  });
});

it('should open Menu Display and focus on last element when pressing Up arrow', () => {
  const wrapper = mount(
    <Menu>
      <MenuButton><a>Toggle Menu</a></MenuButton>
      <MenuDisplay>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    </Menu>
  );

  const menuButton = wrapper.find(MenuButton);
  menuButton.simulate('keydown', { keyCode: 38 });

  expect(wrapper.find(MenuDisplay).length).toEqual(1);
  expect(document.activeElement).toEqual(wrapper.find('h1').get(2));
});

it('should open Menu Display and focus on first element when pressing Down arrow', () => {
  const wrapper = mount(
    <Menu>
      <MenuButton><a>Toggle Menu</a></MenuButton>
      <MenuDisplay>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    </Menu>
  );

  const menuButton = wrapper.find(MenuButton);
  menuButton.simulate('keydown', { keyCode: 40 });

  expect(wrapper.find(MenuDisplay).length).toEqual(1);
  expect(document.activeElement).toEqual(wrapper.find('h1').get(0));
});

it('should focus on correct element when clicking on Menu Button, after previously pressing Up arrow', () => {
  const wrapper = mount(
    <Menu>
      <MenuButton><a>Toggle Menu</a></MenuButton>
      <MenuDisplay>
        <h1>First</h1>
        <h1>Second</h1>
        <h1>Third</h1>
      </MenuDisplay>
    </Menu>
  );
  const menuButton = wrapper.find(MenuButton);
  menuButton.simulate('keydown', { keyCode: 38 });

  menuButton.simulate('click'); // Close the menu display
  menuButton.simulate('click'); // Reopen the menu display

  expect(document.activeElement).toEqual(wrapper.find('h1').get(0));
});
