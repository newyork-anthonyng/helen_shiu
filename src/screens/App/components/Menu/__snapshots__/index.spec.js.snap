// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
<Menu>
  <div>
    <MenuButton
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      <a
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        Toggle Menu
      </a>
    </MenuButton>
  </div>
</Menu>
`;
